### ISSUE TITLE
exit codes

### ISSUE THREAD

**Issue by websafe** [dslackw/slpkg#73]
Hello @dslackw 

I'm building now qemu and virt-manager. Both have some **required**
dependencies, listed in `REQUIRES`, but they also have some **optional**
dependencies, which are **only mentioned in the README**.

These optional dependencies are mostly **autodetected** during build time,
and the build process will continue even if these dependencies are not found
(because they're  optional).

For example: 

[Qemu](https://slackbuilds.org/repository/14.2/system/qemu/) has the following optional dependencies:
 + libcacard
 + spice
 + usbredir
 + virglrenderer
 + device-tree-compiler

If these dependencies are found during build time - `qemu` will be compiled with
libcacard|spice|usbredir|virglrenderer|device-tree-compiler support, and if they're
not found during build time - `qemu` will be built without
libcacard|spice|usbredir|virglrenderer|device-tree-compiler support.

And now my problem. After analyzing all README files I've created a small script,
which should compile `qemu` and `virt-manager` with all optional dependencies
present:

~~~~ bash
#!/bin/bash

#
set -e

#
PACKAGES=$(cat <<EOF
libacard
opus
spice
usbredir
virglrenderer
device-tree-compiler
qemu
avahi
phodav
virt-manager
EOF
)

#
for package in ${PACKAGES};
do
  echo $package
  echo "Y" | slpkg -s sbo ${package}
done
~~~~

The problem is this script will continue even if one of the packages is not found (for example because of a typo):

~~~~ bash
# if echo "Y" | slpkg -s sbo libacard; then echo "OK"; else echo "ERROR"; fi
Reading package lists... Done

Not found packages for installation

OK
~~~~

(should be lib**_c_**acard, not `libacard` below).

So I have to modify my script to redirect build output to a logfile and then check the logfile if the string `Not found packages for installation` was found, which would indicate the package was not found.

Something like this:

~~~~ bash
#!/bin/bash

#
set -e

#
PACKAGES=$(cat <<EOF
libacard
opus
spice
usbredir
virglrenderer
device-tree-compiler
qemu
avahi
phodav
virt-manager
EOF
)

#
for package in ${PACKAGES};
do
  echo $package
  echo "Y" | slpkg -s sbo ${package}  2>&1 >/var/log/build.log
  if grep -q "Not found packages for installation" /var/log/build.log; then
    echo "ERROR"
    exit 1
  fi
done
~~~~

But this makes `set -e` unusable...

So my suggestion is: could `slpkg` return an exit code other than `0` in addition
to `Not found packages for installation` if a package was not found?

This would make things much easier.

I tried to fix this by myself and post a PR, but I'm rather a complete python newbie
and don't want to break things ;-)

**Comment by websafe** [comment]
And where the problem started:

thread: https://lists.slackbuilds.org/pipermail/slackbuilds-users/2016-October/017221.html
 + https://lists.slackbuilds.org/pipermail/slackbuilds-users/2016-October/017224.html `Maybe the *.nfo files should be extended by listing optional
dependencies, so we could handle these an automated way?`
 + https://lists.slackbuilds.org/pipermail/slackbuilds-users/2016-October/017240.html

**Comment by dslackw** [comment]
Return like:
```
slpkg: Error(303): Not found packages for installation
```

???

**Comment by websafe** [comment]
The terminal output is unimportant. But the system exit code is important when using slpkg in scripts or automated build environments.

http://www.tldp.org/LDP/abs/html/exit-status.html
http://stackoverflow.com/questions/285289/exit-codes-in-python
http://stackoverflow.com/questions/6758514/python-exit-codes
http://www.wingware.com/psupport/python-manual/2.7/library/sys.html#sys.exit


Example
------------

File `test-error-zero.sh` contains:
~~~~ bash
#!/bin/bash
exit 0
~~~~

File `test-error-nonzero.sh` contains:
~~~~ bash
#!/bin/bash
exit 1
~~~~

File `check-errors.sh` contains:
~~~~ bash
#!/bin/bash

#
echo -e "\nChecking test-error-zero.sh"
if ./test-error-zero.sh;
then
  echo "TRUE"
else
  echo "FALSE"
fi

#
echo -e "\nChecking test-error-nonzero.sh"
if ./test-error-nonzero.sh;
then
  echo "TRUE"
else
  echo "FALSE"
fi
~~~~


The result of `check-errors.sh`:

~~~~ shell
Checking test-error-zero.sh
TRUE

Checking test-error-nonzero.sh
FALSE
~~~~


In short, exit code `0` means **no error occurred**, and exit code other than `0` means **error occurred**.

**Comment by dslackw** [comment]
I have make little changes in develop branch. Can you try from there?

**Comment by websafe** [comment]
### Stable 3.1.8

Checking on a non-existent package (`libacard` should be `libcacard`):

~~~~
# slpkg --version
Version   : 3.1.8
Licence   : GNU General Public License v3 (GPLv3)
Email     : d.zlatanidis@gmail.com
Maintainer: Dimitris Zlatanidis (dslackw)
bash-4.3# if echo Y | slpkg -s sbo libacard; then echo "OK"; else echo "NOT OK"; fi
Reading package lists... Done

Not found packages for installation

OK
~~~~

## Develop branch

Again, checking with a non-existent `libacard` package, instead of `libcacard`:

~~~~
#  if echo Y | slpkg -s sbo libacard; then echo "OK"; else echo "NOT OK"; fi
Reading package lists... Done

Not found packages for installation

NOT OK
~~~~

So it's working now :-) Thank you @dslackw :-)
Returns exit code `1` on failure.

~~~~
# echo Y | slpkg -s sbo libacard; echo $?
Reading package lists... Done

Not found packages for installation

1
~~~~

Exactly that was missing -  #73 is fixed now in develop branch :-)

Thanks again :-)
